contrib/format-fixedwidth/src/test/java/org/apache/drill/exec/store/fixedwidth/TestFixedwidthRecordReader.java

```diff
@@ -0,0 +1,221 @@
+    dirTestWatcher.copyResourceToRoot(Paths.get("fwf/"));
+  }
+
+  @Test
+  public void testStarQuery() throws Exception {
+    String sql = "SELECT * FROM cp.`fwf/test.fwf`";
+    RowSet results = client.queryBuilder().sql(sql).rowSet();
+    RowSet expected = setupTestData();
+    new RowSetComparison(expected).verifyAndClearAll(results);
+  }
+
+  @Test
+  public void testExplicitAllQuery() throws Exception {
+    String sql = "SELECT Number, Letter, Address, `Date`, `Time`, DateTime FROM cp.`fwf/test.fwf`";
+    RowSet results = client.queryBuilder().sql(sql).rowSet();
+    RowSet expected = setupTestData();
+    new RowSetComparison(expected).verifyAndClearAll(results);
+  }
+
+  @Test
+  public void testExplicitQuery() throws Exception {
+    String sql = "SELECT Number, Letter, Address FROM cp.`fwf/test.fwf` WHERE Letter='yzzz'";
+    QueryBuilder q = client.queryBuilder().sql(sql);
+    RowSet results = q.rowSet();
+
+    TupleMetadata expectedSchema = new SchemaBuilder()
+      .addNullable("Number", TypeProtos.MinorType.VARDECIMAL,38,4)
+      .addNullable("Letter", TypeProtos.MinorType.VARCHAR)
+      .addNullable("Address", TypeProtos.MinorType.INT)
+      .buildSchema();
+    RowSet expected = new RowSetBuilder(client.allocator(), expectedSchema)
+      .addRow(77.77, "yzzz", 777)
+      .build();
+
+    new RowSetComparison(expected).verifyAndClearAll(results);
+  }
+
+  //Test Serialization/Deserialization
+  @Test
+  public void testSerDe() throws Exception {
+    String sql = "SELECT COUNT(*) FROM cp.`fwf/test.fwf`";
+    String plan = queryBuilder().sql(sql).explainJson();
+    long cnt = queryBuilder().physical(plan).singletonLong();
+    assertEquals(25L, cnt);
+  }
+
+  @Test
+  public void testStarQueryWithCompressedFile() throws Exception {
+    generateCompressedFile("fwf/test.fwf", "zip", "fwf/test.fwf.zip" );
+
+    String sql = "SELECT * FROM dfs.`fwf/test.fwf.zip`";
+    RowSet results = client.queryBuilder().sql(sql).rowSet();
+    RowSet expected = setupTestData();
+    new RowSetComparison(expected).verifyAndClearAll(results);
+  }
+
+  // Test Entering invalid schemata - incorrect limits
+    // Undefined field, what happens
+    // Parse invalid file, make sure correct error
+
+
+  @Test
+  public void testOutOfOrder() throws Exception{
+    String sql = "SELECT Address, DateTime, `Date`, Letter FROM cp.`fwf/test.fwf`";
+    QueryBuilder q = client.queryBuilder().sql(sql);
+    RowSet results = q.rowSet();
+
+    TupleMetadata expectedSchema = new SchemaBuilder()
+      .addNullable("Address", TypeProtos.MinorType.INT)
+      .addNullable("DateTime", TypeProtos.MinorType.TIMESTAMP)
+      .addNullable("Date", TypeProtos.MinorType.DATE)
+      .addNullable("Letter", TypeProtos.MinorType.VARCHAR)
+      .buildSchema();
+    RowSet expected = new RowSetBuilder(client.allocator(), expectedSchema)
+      .addRow(567, Instant.parse("2021-02-10T15:30:27.00Z"), LocalDate.parse("2021-02-10"), "test")
```

Comment:
Generally, for something like this, you can use a `LIMIT 2` and only check the first row or two. The code is not going to change column order after the 10th row!